File: frontend/ged/lib/app/data/model/ged_versao_documento_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:ged/app/infra/infra_imports.dart';
import 'package:ged/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';
import 'package:ged/app/data/domain/domain_imports.dart';

class GedVersaoDocumentoModel {
	int? id;
	int? idGedDocumentoDetalhe;
	int? idColaborador;
	String? acao;
	int? versao;
	DateTime? dataVersao;
	String? horaVersao;
	String? hashArquivo;
	String? caminho;
	GedDocumentoDetalheModel? gedDocumentoDetalheModel;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;

	GedVersaoDocumentoModel({
		this.id,
		this.idGedDocumentoDetalhe,
		this.idColaborador,
		this.acao,
		this.versao,
		this.dataVersao,
		this.horaVersao,
		this.hashArquivo,
		this.caminho,
		this.gedDocumentoDetalheModel,
		this.viewPessoaColaboradorModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'acao',
		'versao',
		'data_versao',
		'hora_versao',
		'hash_arquivo',
		'caminho',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Acao',
		'Versao',
		'Data Versao',
		'Hora Versao',
		'Hash Arquivo',
		'Caminho',
	];

	GedVersaoDocumentoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idGedDocumentoDetalhe = jsonData['idGedDocumentoDetalhe'];
		idColaborador = jsonData['idColaborador'];
		acao = GedVersaoDocumentoDomain.getAcao(jsonData['acao']);
		versao = jsonData['versao'];
		dataVersao = jsonData['dataVersao'] != null ? DateTime.tryParse(jsonData['dataVersao']) : null;
		horaVersao = jsonData['horaVersao'];
		hashArquivo = jsonData['hashArquivo'];
		caminho = jsonData['caminho'];
		gedDocumentoDetalheModel = jsonData['gedDocumentoDetalheModel'] == null ? GedDocumentoDetalheModel() : GedDocumentoDetalheModel.fromJson(jsonData['gedDocumentoDetalheModel']);
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idGedDocumentoDetalhe'] = idGedDocumentoDetalhe != 0 ? idGedDocumentoDetalhe : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['acao'] = GedVersaoDocumentoDomain.setAcao(acao);
		jsonData['versao'] = versao;
		jsonData['dataVersao'] = dataVersao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataVersao!) : null;
		jsonData['horaVersao'] = Util.removeMask(horaVersao);
		jsonData['hashArquivo'] = hashArquivo;
		jsonData['caminho'] = caminho;
		jsonData['gedDocumentoDetalheModel'] = gedDocumentoDetalheModel?.toJson;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idGedDocumentoDetalhe = plutoRow.cells['idGedDocumentoDetalhe']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		acao = plutoRow.cells['acao']?.value != '' ? plutoRow.cells['acao']?.value : 'Incluído';
		versao = plutoRow.cells['versao']?.value;
		dataVersao = Util.stringToDate(plutoRow.cells['dataVersao']?.value);
		horaVersao = plutoRow.cells['horaVersao']?.value;
		hashArquivo = plutoRow.cells['hashArquivo']?.value;
		caminho = plutoRow.cells['caminho']?.value;
		gedDocumentoDetalheModel = GedDocumentoDetalheModel();
		gedDocumentoDetalheModel?.nome = plutoRow.cells['gedDocumentoDetalheModel']?.value;
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
	}	

	GedVersaoDocumentoModel clone() {
		return GedVersaoDocumentoModel(
			id: id,
			idGedDocumentoDetalhe: idGedDocumentoDetalhe,
			idColaborador: idColaborador,
			acao: acao,
			versao: versao,
			dataVersao: dataVersao,
			horaVersao: horaVersao,
			hashArquivo: hashArquivo,
			caminho: caminho,
		);			
	}

	
}

File: backend/ged/src/model/ged_versao_documento_model.py
from src import db
from src.model.ged_documento_detalhe_model import GedDocumentoDetalheModel
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel


class GedVersaoDocumentoModel(db.Model):
    __tablename__ = 'ged_versao_documento'

    id = db.Column(db.Integer, primary_key=True)
    acao = db.Column(db.String(1))
    versao = db.Column(db.Integer)
    data_versao = db.Column(db.DateTime)
    hora_versao = db.Column(db.String(8))
    hash_arquivo = db.Column(db.String(250))
    caminho = db.Column(db.String(250))
    id_ged_documento_detalhe = db.Column(db.Integer, db.ForeignKey('ged_documento_detalhe.id'))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))

    ged_documento_detalhe_model = db.relationship('GedDocumentoDetalheModel', foreign_keys=[id_ged_documento_detalhe])
    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_ged_documento_detalhe = data.get('idGedDocumentoDetalhe')
        self.id_colaborador = data.get('idColaborador')
        self.acao = data.get('acao')
        self.versao = data.get('versao')
        self.data_versao = data.get('dataVersao')
        self.hora_versao = data.get('horaVersao')
        self.hash_arquivo = data.get('hashArquivo')
        self.caminho = data.get('caminho')

    def serialize(self):
        return {
            'id': self.id,
            'idGedDocumentoDetalhe': self.id_ged_documento_detalhe,
            'idColaborador': self.id_colaborador,
            'acao': self.acao,
            'versao': self.versao,
            'dataVersao': self.data_versao.isoformat(),
            'horaVersao': self.hora_versao,
            'hashArquivo': self.hash_arquivo,
            'caminho': self.caminho,
            'gedDocumentoDetalheModel': self.ged_documento_detalhe_model.serialize() if self.ged_documento_detalhe_model else None,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
        }